providers/implementations/encode_decode/decode_der2key.c

```diff
@@ -529,19 +563,246 @@ static void ecx_key_adjust(void *key, struct der2key_ctx_st *ctx)
 #  define sm2_d2i_private_key           (d2i_of_void *)d2i_ECPrivateKey
 #  define sm2_d2i_public_key            NULL
 #  define sm2_d2i_key_params            (d2i_of_void *)d2i_ECParameters
+#  define sm2_d2i_PUBKEY                ec_d2i_PUBKEY
+#  define sm2_free                      (free_key_fn *)EC_KEY_free
+#  define sm2_check                     ec_check
+#  define sm2_adjust                    ec_adjust
 
-static void *sm2_d2i_PKCS8(void **key, const unsigned char **der, long der_len,
+static void *sm2_d2i_PKCS8(const unsigned char **der, long der_len,
                            struct der2key_ctx_st *ctx)
 {
     return der2key_decode_p8(der, der_len, ctx,
                              (key_from_pkcs8_t *)ossl_ec_key_from_pkcs8);
 }
-
-#  define sm2_d2i_PUBKEY                (d2i_of_void *)d2i_EC_PUBKEY
-#  define sm2_free                      (free_key_fn *)EC_KEY_free
-#  define sm2_check                     ec_check
-#  define sm2_adjust                    ec_adjust
 # endif
+
+#endif
+
+/* ---------------------------------------------------------------------- */
+
+#ifndef OPENSSL_NO_ML_KEM
+
+/* For lengths between 256 and 65535 the DER encoding is 4 bytes longer */
+#define ML_KEM_OCTET_STRING_OVERHEAD   4
+
+/*-
+ * The DER ASN.1 encoding of ML-KEM and ML-DSA public keys prepends 22 bytes to
+ * the encoded public key:
+ *
+ * - 4 byte outer sequence tag and length
+ * -  2 byte algorithm sequence tag and length
+ * -    2 byte algorithm OID tag and length
+ * -      9 byte algorithm OID (from NIST CSOR OID arc)
+ * -  4 byte bit string tag and length
+ * -    1 bitstring lead byte
+ */
+#define ML_KEM_SPKI_OVERHEAD   22
+
+/*
+ * The accepted private key format is either the 64-bytes (d, z) seed pair, or
+ * else the optionally DER wrapped octet-string encoding  (one byte tag, three
+ * byte length) of the FIPS 203 expanded |dk| key, optionally concatenated with
+ * the public |ek|.  We don't need to worry about whether the public key is or
+ * isn't appended, those are just ignored bytes at the end of an input buffer.
+ */
+typedef enum {
+    NONE_FMT,   /* Unknown or unsupported format */
+    SEED_FMT,   /* Just the seed, no DER wrapping */
+    LONG_FMT,   /* Key(s), no octet-string DER wrapping */
```

Comment:
LONG = raw or encoded

Reply:
Correct, this *could* be DER encoded, I'll fix the comment. Thanks.

Reply:
> LONG = raw or encoded

Actually, not sure what you're trying to say here, the long form is not the seed, it is the NIST FIPS 203 |dk| sans DER wrapping, but optionally with an ignored duplicate |ek| tacked on the end for no obvious reason other than OQS doing it wrong across all the algorithms, for lack of support for PKCS#8 v2.